fendermint/vm/interpreter/src/lib.rs

```diff
@@ -0,0 +1,79 @@
+/// example on the outermost layer the input message can be a mix
+/// of self-contained messages and CIDs proposed for resolution
+/// or execution, while in the innermost layer it's all self-contained.
+/// Some interpreters would act like middlewares to resolve CIDs into
+/// a concrete message.
+///
+/// The execution is asynchronous, so that the middleware is allowed
+/// to potentially interact with the outside world. If this was restricted
+/// to things like scheduling a CID resolution, we could use effects
+/// returned from message processing. However, when a node is catching
+/// up with the chain others have already committed, they have to do the
+/// message resolution synchronously, so it has to be done during
+/// message processing. Alternatively we'd have to split the processing
+/// into async steps to pre-process the message, then synchronous steps
+/// to update the state. But this approach is more flexible, because
+/// the middlewares can decide on a message-by-message basis whether
+/// to forward the message to the inner layer. Unfortunately block-level
+/// pre-processing is not possible, because we are fed the messages
+/// one by one through the ABCI.
+///
+/// There is no separate type for `Error`, only `Output`. The reason
+/// is that we'll be calling high level executors internally that
+/// already have their internal error handling, returning all domain
+/// errors such as `OutOfGas` in their output, and only using the
+/// error case for things that are independent of the message itself,
+/// signalling unexpected problems there's no recovering from and
+/// that should stop the block processing altogether.
+#[async_trait]
+pub trait Interpreter: Sync + Send {
+    type State: Send;
+    type Message: Send;
+    type BeginOutput;
+    type DeliverOutput;
+    type EndOutput;
+
+    /// Called once at the beginning of a block.
+    ///
+    /// This is our chance to to run `cron` jobs for example.
+    async fn begin(&self, state: Self::State) -> anyhow::Result<(Self::State, Self::BeginOutput)>;
+
+    /// Apply a message onto the state.
+    ///
+    /// The state is taken by value, so there's no issue with sharing
+    /// mutable references in futures. The modified value should be
+    /// returned along with the return value.
+    ///
+    /// Only return an error case if something truly unexpected happens
+    /// that should stop message processing altogether; otherwise use
+    /// the output for signalling all execution results.
+    async fn deliver(
+        &self,
+        state: Self::State,
+        msg: Self::Message,
+    ) -> anyhow::Result<(Self::State, Self::DeliverOutput)>;
+
+    /// Called once at the end of a block.
+    ///
+    /// This is where we can apply end-of-epoch processing, for example to process staking
+    /// requests once every 1000 blocks.
+    async fn end(&self, state: Self::State) -> anyhow::Result<(Self::State, Self::EndOutput)>;
```

Comment:
This determines the specific callback to trigger at the end of a block for a specific message, right?

Reply:
I'm not sure what you mean by "for a specific message" :thinking: 

This happens at the end of the block, when all messages have been processed. The concrete use case is that we look at the height of the block (part of `state`) and if, say, `height % 1000 == 0` then we produce a checkpoint. Because it's async, it can even send it if we want, but it can definitely add it to the ledger as well. Another thing that will happen here is that all the staking changes which were buffered on the ledger during the last 1000 blocks are applied, to calculate a new power distribution and return the validator set changes to Tendermint. 

So, there is no specific message at this point, but rather what the ledger state dictates.